modules/light-clients/08-wasm/types/querier.go

```diff
@@ -0,0 +1,150 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"slices"
+
+	wasmvmtypes "github.com/CosmWasm/wasmvm/types"
+
+	errorsmod "cosmossdk.io/errors"
+	storetypes "cosmossdk.io/store/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	abci "github.com/cometbft/cometbft/abci/types"
+
+	"github.com/cosmos/ibc-go/modules/light-clients/08-wasm/internal/ibcwasm"
+)
+
+var (
+	_ wasmvmtypes.Querier = (*DefaultQuerier)(nil)
+
+	queryPlugins = NewDefaultQueryPlugins()
+)
+
+type DefaultQuerier struct {
+	Ctx      sdk.Context
+	CallerID string
+}
+
+// NewDefaultQuerier returns a default querier that can be used in the contract.
+func NewQueryHandler(ctx sdk.Context, callerID string) *DefaultQuerier {
+	return &DefaultQuerier{
+		Ctx:      ctx,
+		CallerID: callerID,
+	}
+}
+
+// GasConsumed implements the wasmvmtypes.Querier interface.
+func (q *DefaultQuerier) GasConsumed() uint64 {
+	return VMGasRegister.ToWasmVMGas(q.Ctx.GasMeter().GasConsumed())
+}
+
+// Query implements the wasmvmtypes.Querier interface.
+func (q *DefaultQuerier) Query(request wasmvmtypes.QueryRequest, gasLimit uint64) ([]byte, error) {
+	sdkGas := VMGasRegister.FromWasmVMGas(gasLimit)
+
+	subCtx, _ := q.Ctx.WithGasMeter(storetypes.NewGasMeter(sdkGas)).CacheContext()
+
+	// make sure we charge the higher level context even on panic
+	defer func() {
+		q.Ctx.GasMeter().ConsumeGas(subCtx.GasMeter().GasConsumed(), "contract sub-query")
+	}()
+
+	if request.Stargate != nil {
+		return GetQueryPlugins().Stargate(subCtx, request.Stargate)
+	}
+
+	if request.Custom != nil {
+		return GetQueryPlugins().Custom(subCtx, request.Custom)
+	}
+
+	return nil, wasmvmtypes.UnsupportedRequest{Kind: "Unsupported query request"}
+}
+
+type (
+	CustomQuerier   func(ctx sdk.Context, request json.RawMessage) ([]byte, error)
+	StargateQuerier func(ctx sdk.Context, request *wasmvmtypes.StargateQuery) ([]byte, error)
+)
+
+// QueryPlugins is a list of query handlers that can be used to extend the default querier.
+type QueryPlugins struct {
+	Custom   CustomQuerier
+	Stargate StargateQuerier
+}
+
+// Merge merges the query plugin with a provided one.
+func (e QueryPlugins) Merge(x *QueryPlugins) QueryPlugins {
+	// only update if this is non-nil and then only set values
+	if x == nil {
+		return e
+	}
+
+	if x.Custom != nil {
+		e.Custom = x.Custom
+	}
+
+	if x.Stargate != nil {
+		e.Stargate = x.Stargate
+	}
+
+	return e
+}
+
+// SetQueryPlugins sets the current query plugins
+func SetQueryPlugins(plugins *QueryPlugins) {
+	queryPlugins = plugins
+}
+
+// GetQueryPlugins returns the current query plugins
+func GetQueryPlugins() *QueryPlugins {
+	return queryPlugins
+}
```

Comment:
If possible it would be nice. There's the option of moving `queryPlugins` to `internal/ibcwasm` that I think @DimitrisJim suggested but unsure if that comes with some import cycle shenanigans.

Reply:
But the keeper needs access to these functions

Reply:
> But the keeper needs access to these functions

The keeper can access these functions when it is internal? The 08-wasm keeper right?

Reply:
> it just wraps things they override (namely, the queryPlugins) hence it could ideally be private (though still in types).

Yea, I think it could be nice to make `queryHandler` and `newQueryHandler` private as it should only be invoked by types/vm.go? But it is a nice to have

Reply:
After reviewing some more, I'd have a preference to move `GetQueryPlugins`/`SetQueryPlugins` to the internal package so that `WithQueryPlugins` is the only option to override the query plugins (ensuring no accidental nil values)

Reply:
Fair enough @colin-axner. Do you think it'd be ok to keep `queryHandler` in types (after it is made private)? So that I don't need to move gas_registry to internal too?

Reply:
yes, I am definitely in favour of keeping all implementation code in types. Make it private when possible. For Get/Set of QueryPlugins, I would add QueryPlugins as an expected interface, so you don't need to move over the implementation logic. Hopefully we remove all this global stuff before too long

Reply:
> Hopefully we remove all this global stuff before too long

Agree! But from the looks of things we should be able to switch over quite nicely from the `internal/ibcwasm` stuff to the `vm` and such existing as fields on the `Keeper`. Looks like it will be clean enough to do it when the time comes :) 

And fwiw I think working with the `internal` dir for the global types makes a big difference for the time being!

Reply:
I moved getters and setters to internal now. And made `queryHandler` private